File: common/src/main/java/juuxel/adorn/client/gui/widget/FlipBook.java
package juuxel.adorn.client.gui.widget;

import net.minecraft.client.gui.components.events.GuiEventListener;

import java.util.ArrayList;
import java.util.List;

public final class FlipBook extends WidgetEnvelope implements PageContainer<GuiEventListener> {
    private final List<GuiEventListener> pages = new ArrayList<>();
    private final Runnable pageUpdateListener;
    private int currentPage = 0;

    public FlipBook(Runnable pageUpdateListener) {
        this.pageUpdateListener = pageUpdateListener;
    }

    @Override
    public int getCurrentPage() {
        return currentPage;
    }

    @Override
    public void setCurrentPage(int currentPage) {
        this.currentPage = currentPage;
        pageUpdateListener.run();
    }

    @Override
    public int getPageCount() {
        return pages.size();
    }

    @Override
    public GuiEventListener getCurrentPageValue() {
        return current();
    }

    @Override
    protected GuiEventListener current() {
        return pages.get(currentPage);
    }

    public void add(GuiEventListener page) {
        pages.add(page);
    }

    @Override
    public void showPreviousPage() {
        if (hasPreviousPage()) {
            currentPage--;
            pageUpdateListener.run();
        }
    }

    @Override
    public void showNextPage() {
        if (hasNextPage()) {
            currentPage++;
            pageUpdateListener.run();
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/gui/widget/FlipBook.kt
package juuxel.adorn.client.gui.widget

import net.minecraft.client.gui.Element

class FlipBook<E : Element> : WidgetEnvelope, PageContainer<E> {
    private val pages = ArrayList<E>()
    override var currentPage = 0
        set(value) {
            field = value
            for (listener in pageUpdateListeners) {
                listener()
            }
        }
    override val pageCount get() = pages.size
    override val currentPageValue get() = current()
    private val pageUpdateListeners: MutableList<() -> Unit> = ArrayList()

    constructor()
    constructor(pageUpdateListener: () -> Unit) : this() {
        pageUpdateListeners += pageUpdateListener
    }

    override fun current() = pages[currentPage]

    fun add(page: E) {
        pages += page
    }

    fun addPageUpdateListener(listener: () -> Unit) {
        pageUpdateListeners += listener
    }

    override fun showPreviousPage() {
        if (hasPreviousPage()) {
            currentPage--
        }
    }

    override fun showNextPage() {
        if (hasNextPage()) {
            currentPage++
        }
    }

    override fun showPage(value: E) {
        val index = pages.indexOf(value)
        if (index >= 0) {
            currentPage = index
        }
    }
}
